ENH: C/NOFS IVM Cleaning Routine Update

jklenzing opened the pull request:
# Description

Addresses #115.

Currently the C/NOFS IVM cleaning routine only provides data for where the drifts are reliable.  However, it is possible in some cases to have reliable density, composition, and temperature information where the drifts are suspect.  This updates the clean parameters to allow all desired data through while replacing suspect drifts with NaNs.

Instrument quality flags described in https://cdaweb.gsfc.nasa.gov/pub/data/cnofs/cindi/CINDI%20IVM%20Quality%20Flag%20Description.doc (last updated 2018-01-25).

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

tested ivm.load locally.

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

jklenzing commented:
Hmm.  Slight decrease in coverage due to not covering final `else` statement.

https://coveralls.io/builds/25831087/source?filename=pysat%2Finstruments%2Fcnofs_ivm.py#L127

Shouldn't testing clean_level=None wind up here?

aburrell commented on pysat/instruments/cnofs_ivm.py:
```diff
@@ -113,31 +113,39 @@ def clean(inst):
     # make sure all -999999 values are NaN
```
The altitude limit for clean should be removed

rstoneback commented:
If clean level is set to None the clean method shouldn’t be called. That’s what I remember at least. On phone.

jklenzing commented:
> If clean level is set to None the clean method shouldn’t be called. That’s what I remember at least. On phone.

Makes sense.

jklenzing commented:
Fixed the 'dirty' vs None case, removed altitude limits.  Will produce some examples of data for the record.  I'm open to adding the density and composition limits, but we should work out the exact limits first.

From my Mar 5 notes in the issue:
- Ion temperatures should require a O^+ fraction of 0.15 or higher, citing the Hairston paper.
- From my notes, drifts should be based on a combination of [O^+] and flag information. In my old matlab routines, there is a 'dmpts' parameter in addition to the drift meter flag which I don't see in the pysat metadata. It's likely my shorthand for something rather than an official name.
- I would also suggest dropping the drifts for 100% O^+.

The first and third are straightforward to implement.  The second requires some more work.

aburrell commented:
So I've been digging through my C code looking for the different quality controls I used in my dissertation.  Here is the 'RPA is negligible' allowance:

```
/* bad == 0 if there is no sunglint */
 if(ivm_data->qual_flag == 1 && bad == 0)
{
  if(data_ptr->rpa_quality_flag == 3.0 &&
     data_ptr->idens * data_ptr->imassfrac[0] <= 3.0e4)
    {
      /* Test and see if the conditions allow for a data allowance */
      k = 3;

      if(ivm_data->vp_flag < 2 || p_perc < .01) k--;
      if(ivm_data->vm_flag < 2 || m_perc < .01) k--;
      if(ivm_data->vz_flag < 2 || z_perc < .01) k--;

      if(k == 0 && (ivm_data->vp_flag == 2 || ivm_data->vm_flag == 2
		    || ivm_data->vz_flag == 2))
	{
	  printf("%s ADVISEMENT: ram velocity ", module_name);
	  printf("discarded at row [%d/%d]\n", n,
	  	 temp_data->number_samples);
	  data_ptr->vx = 0.0;
	}
      else
	{
	  printf("%s ADVISEMENT: RPA flag triggered at row [%d/%d]\n",
		 module_name, n, temp_data->number_samples);
	  bad = 1;
	}
    }

  if(bad == 0 && data_ptr->idm_quality_num < 9)
    {
      printf("%s ADVISEMENT: IDM flag triggered in row [%d/%d]\n",
	     module_name, n, temp_data->number_samples);
      bad = 1;
    }
}
```

jklenzing commented:
Currently we have:
- Clean has RPAflag <=1
- Dusty has RPAflag <=3

The drifts in question (if I understand the above code) are those where RPAflag is 3 and ([O+] > 3e4 or X_perc < 0.01)

Should we:
- Open up `clean` to include these drifts?
- Restrict `dusty` to ignore drifts where this is not true?

aburrell commented:
Well, I used those drifts in all my studies, so I think they are good.  And they are certainly cleaner than the current dusty level.

jklenzing commented:
OK, as discussed in the meeting, the additional changes based on plasma parameters are now only run for the 'dusty' levels.  Simplifies and streamlines the code.

Also, additional ion temperature cleaning now for 'clean' and 'dusty' levels only.

Ready for new review.

aburrell commented:
I would test the implementation and create a plot to be sure, but pysatCDF isn't working for me, so I can't load CINDI data :-(

jklenzing commented:
I will add to the docstring. I don’t think the 100% O+ case is published yet. @rstoneback?

rstoneback commented:
Nope, no publications on the ultra high O+ case.
